src/transformStateWithClones.js

```diff
@@ -7,7 +7,27 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const result = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const keys of action.keysToRemove) {
+          delete stateCopy[keys];
+        }
+        break;
 
+      default:
+        stateCopy = {};
+    }
```

Comment:
Consider adding a 'clear' case to the switch statement. This will make it explicit how the 'clear' action should be handled and will improve code readability. The default case should be used for error handling, e.g., when an action type is not recognized.